### PR TITLE
Pinned answers

### LINKED ISSUE #955: Pinned answers
Some people don't want to answer particular things repeatedly, so something akin to a FAQ section would be nice. We could solve this by adding the ability to pin answers (like most microblog social networks offer nowadays) that these are either pinned at the top of the "Answers" tab on profiles, or get their own "Pinned" tab.

### PR DESCRIPTION
**Pull request by raccube**
Closes #955

### CODE REVIEW

File: lib/use_case/answer/unpin.rb [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+          status:   200,
+          resource: nil,
+        }
+      end
+
+      private
+
+      def check_ownership!
+        raise ::Errors::NotAuthorized unless answer.user == user
+      end
```
Comment (pixeldesu) [inline lib/use_case/answer/unpin.rb]: Do we need custom logic for this when we have Pundit in the project?

**Review by pixeldesu — CHANGES_REQUESTED** [review]
See comment.

Also, I'm not sure I like the approach with two render sets of answers. In the worst case it means initially rendering 10+page size answers, instead of just the pinned ones.

Is there no way to make/combine a query that shows the pinned entries first?

File: app/views/actions/_pin.html.haml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+- if answer.pinned?
+  = button_to unpin_answer_path(id: answer.id),
+    class: "dropdown-item",
+    method: :delete,
+    form: { id: "ab-pin-#{answer.id}", data: { turbo_stream: true } } do
+    %i.fa.fa-fw.fa-thumbtack
+    = t(".unpin")
+- else
+  = button_to pin_answer_path(id: answer.id),
```
Comment (nilsding) [inline app/views/actions/_pin.html.haml]: i get a

```
ActionController::UrlGenerationError - No route matches {:action=>"pin", :controller=>"answer", :id=>109842029517652183}, missing required keys: [:username]
Did you mean?  pin_answer_url
               unpin_answer_url
               unpin_answer_path:
  app/views/actions/_pin.html.haml:9
  app/views/actions/_answer.html.haml:20
  app/views/answerbox/_actions.html.haml:16
  app/views/application/_answerbox.html.haml:24
  app/views/timeline/timeline.html.haml:3
  app/views/timeline/timeline.html.haml:2
  app/controllers/timeline_controller.rb:39:in `block (2 levels) in paginate_timeline'
  app/controllers/timeline_controller.rb:38:in `paginate_timeline'
  app/controllers/timeline_controller.rb:9:in `index'
  app/controllers/application_controller.rb:23:in `switch_locale'
```

when trying to view my timeline and having a pinned answer

**Review by nilsding — CHANGES_REQUESTED** [review]

File: app/views/actions/_pin.html.haml (hunk 2 above)
Comment (raccube) [reply app/views/actions/_pin.html.haml]: c3e48b7

File: spec/controllers/answer_controller_spec.rb [hunk 3]
```diff
@@ -39,4 +41,62 @@
...
+
+    context "other user signed in" do
+      let(:other_user) { FactoryBot.create(:user) }
+
+      before(:each) do
+        sign_in other_user
+        answer.update!(pinned_at: Time.at(1603290950).utc)
+      end
+
+      it "does not pin the answer do" do
```
Comment (nilsding) [inline spec/controllers/answer_controller_spec.rb]:
```suggestion
      it "does not unpin the answer" do
```

File: spec/controllers/answer_controller_spec.rb [hunk 4]
```diff
@@ -39,4 +41,62 @@
...
+  describe "#unpin" do
+    subject { delete :unpin, params: { username: user.screen_name, id: answer.id } }
+
+    context "user signed in" do
+      before(:each) do
+        sign_in user
+        answer.update!(pinned_at: Time.at(1603290950).utc)
+      end
+
+      it "pins the answer" do
```
Comment (nilsding) [inline spec/controllers/answer_controller_spec.rb]:
```suggestion
      it "unpins the answer" do
```

File: spec/controllers/answer_controller_spec.rb [hunk 5]
```diff
@@ -39,4 +41,62 @@
...
+        end
+      end
+    end
+
+    context "other user signed in" do
+      let(:other_user) { FactoryBot.create(:user) }
+
+      before(:each) { sign_in other_user }
+
+      it "does not pin the answer do" do
```
Comment (nilsding) [inline spec/controllers/answer_controller_spec.rb]:
```suggestion
      it "does not pin the answer" do
```

**Review by nilsding — APPROVED** [review]
seems to work nice overall; would be cool to have a notification indicating success/failure though.

**Review by pixeldesu — APPROVED** [review]

**Review by pixeldesu — APPROVED** [review]

**Pull request merged**